packages/devtools_app/lib/src/debugger/debugger_controller.dart

```diff
@@ -624,31 +627,24 @@ class DebuggerController extends DisposableController
 
     // Perform an early exit if we're not paused.
     if (!paused) {
-      _populateFrameInfo([]);
+      _populateFrameInfo([], truncated: false);
       return;
     }
 
-    // First, notify based on the single 'pauseEvent.topFrame' frame.
-    if (pauseEvent?.topFrame != null) {
-      final tempFrames = _framesForCallStack(
-        [pauseEvent.topFrame],
-        reportedException: pauseEvent?.exception,
-      );
-      _populateFrameInfo(
-        [await _createStackFrameWithLocation(tempFrames.first)],
-        truncated: true,
-      );
-      _log.log('created first frame');
-    }
-
-    // We populate the first 10 frames to match the behavior in VS Code.
+    // We populate the first 12 frames; this ~roughly corresponds to the number
+    // of visible stack frames.
     _getStackOperation =
-        CancelableOperation.fromFuture(_getStackInfo(limit: 10));
+        CancelableOperation.fromFuture(_getStackInfo(limit: 12));
     final stackInfo = await _getStackOperation.value;
     _populateFrameInfo(
       stackInfo.frames,
       truncated: stackInfo.truncated ?? false,
     );
+
+    // In the background, populate the rest of the frames.
+    if (stackInfo.truncated) {
+      unawaited(_getFullStack());
```

Comment:
I think we need to rate limited requests to _getFullStack. You can use the RateLimiter class to make sure that only 1 request is running at a time. Otherwise if you step quickly you might end up with 20 pending stack trace requests which could bog down flutter web. Would be nice if the VM Service had a way for us to cancel requests or detect that a getFullStack request should be ignored if execution resumed since when the request was issued.

Reply:
If we rate limit the requests here, we wouldn't be sending the request for the latest (current) set of frames? I'm trying to think of how we'd actually use this.

Reply:
Here is how you do what you want with `RateLimiter`. It is more of a scheduler than a true RateLimiter as it doesn't ignore requests, it just ensures that the rate limit is respected.
We should probably add a wrapper class on top of RateLimiter that adds the extra logic to toss results that are known to be stale as a newer result was requested.
```dart
int nextTimeStampRequestId = 0;
myRateLimiter = RateLimiter(1.0, () async {
  final requestId = nextTimeStampRequestId++;
  var stackTrace = await fetchStackTrace();
  if (requestId != nextTimeStampRequestId) {
    // Stale request... ignore. Also ignore if we are no longer paused, etc.
    return;
  }
  // show the StackTrace in the ui.
  stackTraceValueNotifier.value = stackTrace;
}

call
stackTrace. scheduleRequest() any time you need a full stacktrace for a new frame.  The rate limiter will ensure you don't have more than one simultaneous request and you don't violate the rate limiters frame rate.
```

Reply:
I think I'd prefer to land the PR as is, w/o adding additional code to rate limit requests, and instead work on making populating frame info more lazy. Still, happy to add some rate limiting here if you think it's net useful.

Reply:
Note that in `package:dwds` we ensure the expensive work is only done once so rate limiting is not super necessary for the web workflow.

Reply:
do you cancel pending requests in dwds if they become obsolete (e.g. stack frame changed)? that would mean this doesn't need to be rate limited.

Reply:
We do not cancel requests. That could present a problem.

Reply:
I think we should just move to requesting all frames, and requesting them with a parameter so the variables aren't populated. There can then be another method (getObject(), getStack(), ...), to request more info for a specific frame. That should make everything fast and keep various implementations ~simple.